### ISSUE TITLE
[PW-2483]: Adyen module should not try to create a credit memo for a refund that already has a credit memo (as refund was initialized in Magento).

### ISSUE THREAD

**Issue by aloupfor** [Adyen/adyen-magento2#722]
Magento Plugin platform version 2.3.2-p2
Plugin version 5.4.0

**Describe the bug**
Adyen module tries to create a credit memo for a refund that already has a credit memo (as refund was initialized in Magento).

**To Reproduce**
Step 1: I create a credit memo on Magento with an online refund for an order with an immediate capture payment (SoFort).
Step 2: Refund is handled by Adyen
Step 3: Adyen then send a notification with `eventCode: REFUND`.
Step 4: Adyen module will try to create a new credit memo because of the code: https://github.com/Adyen/adyen-magento2/blob/5.4.0/Model/Cron.php#L1196-L1201

This steps result in:

```
May 12, 2020 11:34:31 AM Partially Refund Customer Not Notified
IPN "Refunded". Refund issued by merchant. Registered notification about refunded amount of €94.90.
Transaction ID: "". Credit Memo has not been created. Please create offline Credit Memo.
```

in "Notes for this Order" in Magento (on partial refund).

**Expected behavior**
Adyen module should not try to create a credit memo for a refund that already has a credit memo (as refund was initialized in Magento).

**Additional context**
I encountered this behavior with payment methods that do immediate capture.

I wrote the following code (triggered by a custom import of canceled items) to handle cancellation of item that are already invoiced:

```
        // Items that have been invoiced and paid need to be "Refunded" instead of "Canceled".
        // Items from order with immediate capture payment method are invoiced immediately.
        //
        // See https://docs.adyen.com/plugins/magento-2/set-up-point-of-sale#step-2-set-up-capture-delay
        // The capture delay setting does not affect Alipay, WeChatPay, and some other wallets,
        // as these payment methods are always completed in one step.
        /** @var Invoice $invoice */
        $invoice = $order->getInvoiceCollection()->getFirstItem();
        if ($invoice->canRefund()) {
            // See \Magento\Sales\Controller\Adminhtml\Order\CreditmemoLoader::load
            $qtys = [];
            foreach ($orderItems as $orderItem) {
                $qtys[$orderItem->getItemId()] = $items[$orderItem->getSku()];
            }
            $data['qtys'] = $qtys;
            $creditmemo = $this->creditmemoFactory->createByInvoice($invoice, $data);

            // See \Magento\Sales\Controller\Adminhtml\Order\Creditmemo\Save::execute
            $this->creditmemoManagement->refund($creditmemo);

            // Return early as items will be "Refunded"
            // so they do not need to be "Canceled"
            return;
        }
```

I did a test.
On Adyen side, I see:

```
Received 2020-05-06 18:28:27 CEST
Authorised 2020-05-06 18:28:27 CEST
SentForSettle 2020-05-06 18:28:27 CEST
Settled 2020-05-07 02:15:38 CEST
SentForRefund 2020-05-08 18:35:09 CEST
Refunded 2020-05-08 18:35:09 CEST
```

On Magento side ("Notes for this Order"), I see:

```
May 12, 2020 11:34:31 AM Partially Refund Customer Not Notified
Adyen Refund Successfully completed
May 12, 2020 11:34:31 AM Partially Refund Customer Not Notified
IPN "Refunded". Refund issued by merchant. Registered notification about refunded amount of €94.90. Transaction ID: "". Credit Memo has not been created. Please create offline Credit Memo.

May 12, 2020 11:34:31 AM Partially Refund Customer Not Notified
Adyen HTTP Notification(s):
eventCode: REFUND
pspReference: 851588782800945J
paymentMethod: directEbanking
success: true

May 6, 2020 6:33:20 PM Authorized Customer Notified
We refunded €94.90 online. Transaction ID: "881588782460670A-refund"

May 6, 2020 6:33:20 PM Authorized Customer Not Notified
Adyen Result response:
authResult: [refund-received]
pspReference: 851588782800945J

May 6, 2020 6:30:20 PM Authorized Customer Not Notified
Adyen Payment Successfully completed

May 6, 2020 6:30:20 PM Authorized Customer Not Notified
Payment is authorised waiting for capture

May 6, 2020 6:30:20 PM Pending Customer Not Notified
Adyen HTTP Notification(s):
eventCode: AUTHORISATION
pspReference: 881588782460670A
paymentMethod: directEbanking
success: true

May 6, 2020 6:28:16 PM Customer Not Notified
Adyen Result URL response:
authResult: Authorised
pspReference: 881588782460670A
paymentMethod: directEbanking

May 6, 2020 6:27:38 PM Pending Customer Notified

May 6, 2020 6:27:38 PM Customer Not Notified
Adyen Result response:
authResult: RedirectShopper
```

I was worried about the message:
```
IPN "Refunded". Refund issued by merchant. Registered notification about refunded amount of €94.90.
Transaction ID: "". Credit Memo has not been created. Please create offline Credit Memo.
```

From further investigations it looks like this message seems to be added (only on partial refund) because Adyen module did not success to determine that it should not try to create a credit memo with code :
https://github.com/Adyen/adyen-magento2/blob/5.4.0/Model/Cron.php#L1196-L1201

I guess that on processing REFUND notification, Adyen module compared 881588782460670A-refund (`$lastTransactionId`) with 851588782800945J (`$this->_pspReference`).
It then tried to create a credit memo (as those values are different) but it should not as refund is initialize in Magento and credit memo already exists.

---

What can be done to avoid that Adyen module tries to create a credit memo for a refund that already has a credit memo (as refund was initialized in Magento) ?

Is it an issue that is fixed in latest module version ?
It seems to me that there is no change that resolve this problem in the latest version.
I don't see anything related to the issue in : https://github.com/Adyen/adyen-magento2/releases in features, fixes, fixed issues list.
I don't want to upgrade to the latest version to discover that the issue still occurs.

**Comment by cyattilakiss** [comment]
Hi @aloupfor ,

Thank you for raising the issue. We are going to try to reproduce the issue based on your very detailed ticket - highly appreciated!
We haven't touched this part in the latest release so I would recommend you to wait for our response here before upgrading.

Best,
Attila
Adyen

**Comment by peterojo** [comment]
This issue has been fixed in #1064 
Closing this issue now
